test/studies/hpcc/common/bradc/BradsBlock1D.chpl

```diff
@@ -38,18 +38,24 @@ class Block1DDist {
 
   //
   // an associative array of local distribution class descriptors --
-  // set up in initialize() below
+  // set up in initializer below
   //
   // TODO: would like this to be const and initialize in-place,
-  // removing the initialize method
+  // removing the initializer
   //
   // TODO: Remove second generic parameter -- seems confusing and wrong.  Replace
   // with explicit typing of locid field in LocBlock1DDist class.  Particularly
   // since I'm passing in a value from 0.. rather than an actual index value.
   //
   var locDist: [targetLocDom] LocBlock1DDist(glbIdxType, index(targetLocs.domain));
 
-  proc initialize() {
+  proc init(type idxType = int(64), bbox, targetLocs,
+            targetLocDom = targetLocs) {
+    glbIdxType = idxType;
+    this.bbox = bbox;
+    this.targetLocs = targetLocs;
+    this.targetLocDom = targetLocDom;
+    super.init();
```

Comment:
This seems like the kind of case where being able to supply a postInit()-style hook would be beneficial—haven't you essentially written the compiler-provided initializer explicitly here?

Patterns like this are the kind that I would propose putting off rewriting pending further decisions regarding enabling users to leverage the compiler-provided initializer or create a postInt() hook.

Reply:
It's missing the last field.

This one only got called with two arguments, but it seemed like we would probably want the other two in some situations (otherwise the targetLocDom would always match targetLocs, and the index type would always be int).

I'm not bothered by rewriting four fields, personally.  5+ is where I draw the line.  But I can revert this change if you prefer.

Reply:
I don't think the point should be how the initializer is called in practice in the test system (particularly for code that is library-like for which the test may not show all calls that a client of the library might make); nor how many fields you/the author would have to explicitly initialize.  Instead, I think we should focus on identifying cases where the compiler-provided initializer was being leveraged by initialize() in a way that permitted the user to avoid rewriting what the compiler provided (or even 80% of it) themselves.

I'm OK with this change going in, I'm just not comfortable with writing it off as though it was not an interesting use case for initialize().

Reply:
Sorry, I'm using these metrics as short hand for the larger point.  Let me go into a little more detail on what I mean.

> I don't think the point should be how the initializer is called in practice in the test system

In normal cases, I think having a single call style for the creation of the type indicates either a short-sightedness on part of the test author (didn't realize that their type could be initialized without their preprogrammed field relationships), or that we need more unit tests on the type to ensure proper behavior.

Having the initialize method as a fall back enables a lazy response to being confronted with this case, which is why I called it out in my commit message rather than sweeping it entirely under the rug.

> (particularly for code that is library-like for which the test may not show all calls that a client of the library might make);

That is very reasonable.

> nor how many fields you/the author would have to explicitly initialize.

I think we're saying the same thing here: if the amount of complexity needed is sufficient to warrant more than just the default initializer, there is a particular problem space for "I want a post-hook" versus "I will gladly write my own initializer".  The number I gave should be thought of more as the number of fields that aren't impacted by the code you want to add - if it's small, I wouldn't even consider an initialize method myself.

Reply:
> The number I gave should be thought of more as the number of fields that aren't impacted by the code you want to add - if it's small, I wouldn't even consider an initialize method myself.

Err, it's kinda an interplay between percentage and number of fields?  The default initializer being sufficient for 50% of fields is one thing when there are 14 fields versus 2.